### ISSUE TITLE
[FR] Remove useless keywords on `final` classes

### ISSUE THREAD

**Issue by SpacePossum** [FriendsOfPHP/PHP-CS-Fixer#1899]
I would like to write and add a fixer that removes the `final` keyword from methods of a `final` declared class. ~~Secondly on a `final` class `protected` methods can safely be made `private` I think.~~

**Comment by gharlan** [comment]
> Secondly on a final class protected methods can safely be made private I think.

No: https://3v4l.org/GNsrP

**Comment by SpacePossum** [comment]
check, thanks :+1:

**Comment by Soullivaneuh** [comment]
Looks good. We can also remove `final` keyword from `private` method.

A discussion about this was made in Sonata group: https://github.com/sonata-project/dev-kit/issues/185

We find that a `private` method can be "overridden". I said "overridden" because it's not really an override, just the possibility to have the same name. But it was broken if method signatures differ between the child and the parent.

At the end, we supposed it's a PHP bug, and this one was [confirmed](https://bugs.php.net/bug.php?id=72496) and [patched](https://github.com/php/php-src/pull/1980).

So we can say adding the `final` keyword to a private method is completely useless and should be removed.

I think this have a place to this fixer, what do you think?

**Comment by SpacePossum** [comment]
```php
<?php
class Foo
{
    final private function test()
    {
        
    }
}

final class Bar extends Foo
{
    final private function test()
    {
        
    }
}
```

https://3v4l.org/6vgbs

If I get it right, it should be allowed to "override" a final private method after the PR you pointed out is merged and new PHP versions released?
If so I'm happy to see the `final` from private methods removed by a fixer, but not until the new PR versions are out so we can test against this (since before that the code is invalid).

**Comment by Soullivaneuh** [comment]
> If I get it right, it should be allowed to "override" a final private method after the PR you pointed out is merged and new PHP versions released?

Should be. They don't tested with final but about method signature change (should not be an impact).

But yes it should be allowed.

> If so I'm happy to see the `final` from private methods removed by a fixer, but not until the new PR versions are out so we can test against this (since before that the code is invalid).

You're right. It should be out on PHP 5.6.24 (RC at the moment): https://github.com/php/php-src/commit/08777e9615868cf2acf4dc2a86a7ef21106beddd

**Comment by SpacePossum** [comment]
I'm closing this as I use a diff. system for keeping tracking on stuff I want to do one day when I've time. Please don't reopen, I like to have a small list of issues on GH to manage. Thanks!
